flow/BlockCipher.cpp

```diff
@@ -0,0 +1,388 @@
+/*
+ * BlockCipher.cpp
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "flow/BlockCipher.h"
+#include "flow/Error.h"
+#include "flow/FastRef.h"
+#include "flow/IRandom.h"
+#include "flow/network.h"
+#include "flow/Trace.h"
+#include "flow/UnitTest.h"
+#include <cstring>
+#include <memory>
+
+#if ENCRYPTION_ENABLED
+
+uint64_t BlockCipherKeyCache::CIPHER_KEY_CACHE_TTL_NS = 10 * 60 * 1000 * 1000 * 1000L;
+uint8_t EncryptBlockCipher::ENCRYPT_HEADER_VERSION = 1;
+
+// BlockCipherEncryptHeader
+BlockCipherEncryptHeader::BlockCipherEncryptHeader() {
+	flags.headerVersion = EncryptBlockCipher::ENCRYPT_HEADER_VERSION;
+	tag = std::make_unique<uint8_t[]>(AES_256_TAG_LENGTH);
+}
+
+// BlockCipherKey class methods
+
+BlockCipherKey::BlockCipherKey(const BlockCipherDomainId& domainId,
+                               const BlockCipherBaseKeyId& baseCiphId,
+                               const uint8_t* baseCiph,
+                               int baseCiphLen) {
+	BlockCipherRandomSalt salt;
+	if (g_network->isSimulated()) {
+		salt = deterministicRandom()->randomUInt64();
+	} else {
+		salt = nondeterministicRandom()->randomUInt64();
```

Comment:
Why use non-deterministic random outside simulation? Usually we just use the same random number generation in either case.

Reply:
Deterministic random numbers are considered as 'vulnerability' from the code security perspective; it came out as an action item talking to DB-Sec review.

Reply:
looks like this is actually deterministic random? The only implementation I can find is here: https://github.com/apple/foundationdb/blob/ec64890ac1eab074b2712e2613f25aea5ca0273f/flow/flow.cpp#L116 For crypto related randomness, I think we should use /dev/urandom